File: swift/Tennis/TennisGame2.swift
import Foundation

struct Player {
    let name: String
    var point: Int
}

class TennisGame2: TennisGame {
    private var playerOne: Player
    private var playerTwo: Player

    required init(player1: String, player2: String) {
        playerOne = Player(name: player1, point: 0)
        playerTwo = Player(name: player2, point: 0)
    }
    
    func getScore(point1: Int, point2: Int) -> String {
        var score = ""
        switch (point1) {
        case 0: score = "Love"
        case 1: score = "Fifteen"
        case 2: score = "Thirty"
        default: break
        }
        return  "\(score)-All"
    }
    
    func getPoint(_ point: Int) -> String {
        var score = ""
        if (point==1)
        { score = "Fifteen" }
        if (point==2)
        { score = "Thirty" }
        if (point==3)
        { score = "Forty" }
        if (point==4)
        { score = "Love" }
        
        return score
    }
    
    var score: String? {
        var score = ""
        var P1res: String = ""
        var P2res: String = ""
        
        if playerOne.point == playerTwo.point && playerOne.point < 3 {
           score = getScore(point1: playerOne.point, point2: playerTwo.point)
        }
        if playerOne.point==playerTwo.point && playerOne.point>2
        { score = "Deuce" }

        if playerOne.point > 0 && playerTwo.point==0
        {
            P1res = getPoint(playerOne.point)
            P2res = "Love"
            score = "\(P1res)-Love"
        }
        
        if playerTwo.point > 0 && playerOne.point==0
        {
            P2res = getPoint(playerTwo.point)
            P1res = "Love"
            score = "\(P1res)-\(P2res)"
        }
        
        if (playerOne.point>playerTwo.point && playerOne.point < 4)
        {
            P1res = getPoint(playerOne.point)
            if (playerTwo.point==1 || playerTwo.point==2) {
                P2res = getPoint(playerTwo.point)
            }
            score = "\(P1res)-\(P2res)"
        }
        
        if playerTwo.point>playerOne.point && playerTwo.point < 4 {
            P2res = getPoint(playerTwo.point)
            if (playerOne.point==1 || playerOne.point==2) {
                P1res = getPoint(playerOne.point)
            }
            score = "\(P1res)-\(P2res)"
        }
        
        if let newScore = advantageorWinForPlayer() {
            score = newScore
        }
        
        return score
    }
//TODO: Make this functional, eventually (i.e. pass in the player objects)
    func advantageorWinForPlayer() -> String? {
        var score: String?
        
        if playerOne.point > playerTwo.point && playerTwo.point >= 3
        {
            score = "Advantage " + playerOne.name
        }
        
        if playerTwo.point > playerOne.point && playerOne.point >= 3
        {
            score = "Advantage " + playerTwo.name
        }
        
        if playerOne.point>=4 && playerTwo.point>=0 && (playerOne.point-playerTwo.point)>=2
        {
            score = "Win for " + playerOne.name
        }
        if playerTwo.point>=4 && playerOne.point>=0 && (playerTwo.point-playerOne.point)>=2
        {
            score = "Win for " + playerTwo.name
        }
         return score
    }
    
   private func setP1Score(number: Int) {
        
        for _ in 0..<number {
            playerOne.point+=1
        }
        
    }

    private func setP2Score(number: Int) {

        for _ in 0..<number {
            playerTwo.point+=1
        }
        
    }
    
  
    func wonPoint(_ playerName: String) {
        //playerOne.wonPoint(playerName)
        //playerTwo.wonPoint(playerName)
        if playerName == "player1" {
            playerOne.point+=1
        } else {
            playerTwo.point+=1
        }
    }
}

File: java/src/main/java/TennisGame2.java

class Player
{

    public int Point;
}
public class TennisGame2 implements TennisGame
{
    public int P1point = 0;
    public int P2point = 0;
    
    public String P1res = "";
    public String P2res = "";
    private String player1Name;
    private String player2Name;

    private Player Player1 = new Player();

    public TennisGame2(String player1Name, String player2Name) {
        this.player1Name = player1Name;
        this.player2Name = player2Name;
    }

    public String getScore(){
        String score = "";
        if (Player1.Point
                == P2point && Player1.Point
                < 4)
        {
            if (Player1.Point
                    ==0)
                score = "Love";
            if (Player1.Point
                    ==1)
                score = "Fifteen";
            if (Player1.Point
                    ==2)
                score = "Thirty";
            score += "-All";
        }
        if (Player1.Point
                ==P2point && Player1.Point
                >=3)
            score = "Deuce";
        
        if (Player1.Point
                > 0 && P2point==0)
        {
            if (Player1.Point
                    ==1)
                P1res = "Fifteen";
            if (Player1.Point
                    ==2)
                P1res = "Thirty";
            if (Player1.Point
                    ==3)
                P1res = "Forty";
            
            P2res = "Love";
            score = P1res + "-" + P2res;
        }
        if (P2point > 0 && Player1.Point
                ==0)
        {
            if (P2point==1)
                P2res = "Fifteen";
            if (P2point==2)
                P2res = "Thirty";
            if (P2point==3)
                P2res = "Forty";
            
            P1res = "Love";
            score = P1res + "-" + P2res;
        }
        
        if (Player1.Point
                >P2point && Player1.Point
                < 4)
        {
            if (Player1.Point
                    ==2)
                P1res="Thirty";
            if (Player1.Point
                    ==3)
                P1res="Forty";
            if (P2point==1)
                P2res="Fifteen";
            if (P2point==2)
                P2res="Thirty";
            score = P1res + "-" + P2res;
        }
        if (P2point>Player1.Point
                && P2point < 4)
        {
            if (P2point==2)
                P2res="Thirty";
            if (P2point==3)
                P2res="Forty";
            if (Player1.Point
                    ==1)
                P1res="Fifteen";
            if (Player1.Point
                    ==2)
                P1res="Thirty";
            score = P1res + "-" + P2res;
        }
        
        if (Player1.Point
                > P2point && P2point >= 3)
        {
            score = "Advantage player1";
        }
        
        if (P2point > Player1.Point
                && Player1.Point
                >= 3)
        {
            score = "Advantage player2";
        }
        
        if (Player1.Point
                >=4 && P2point>=0 && (Player1.Point
                -P2point)>=2)
        {
            score = "Win for player1";
        }
        if (P2point>=4 && Player1.Point
                >=0 && (P2point-Player1.Point
        )>=2)
        {
            score = "Win for player2";
        }
        return score;
    }
    
    public void SetP1Score(int number){
        
        for (int i = 0; i < number; i++)
        {
            P1Score();
        }
            
    }
    
    public void SetP2Score(int number){
        
        for (int i = 0; i < number; i++)
        {
            P2Score();
        }
            
    }
    
    public void P1Score(){

        Player1.Point++;
    }
    
    public void P2Score(){
        P2point++;
    }

    public void wonPoint(String player) {
        if (player == "player1")
            P1Score();
        else
            P2Score();
    }
}